File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/additive/AdditiveTest.java
package openfoodfacts.github.scrachx.openfood.models.entities.additive;

import org.greenrobot.greendao.DaoException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import openfoodfacts.github.scrachx.openfood.models.DaoSession;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Tests for {@link Additive}
 */
@RunWith(MockitoJUnitRunner.class)
public class AdditiveTest {
    private static final String ADDITIVE_NAME_NAME_1 = "AdditiveName";
    private static final String ADDITIVE_NAME_NAME_2 = "AdditiveName2";
    private static final AdditiveName ADDITIVE_NAME_1 = new AdditiveName(ADDITIVE_NAME_NAME_1);
    private static final AdditiveName ADDITIVE_NAME_2 = new AdditiveName(ADDITIVE_NAME_NAME_2);
    @Mock
    private DaoSession mockDaoSession;
    @Mock
    private AdditiveDao mockAdditiveDao;
    @Mock
    private AdditiveNameDao mockAdditiveNameDao;
    private Additive mAdditive;

    @Before
    public void setup() {
        when(mockDaoSession.getAdditiveNameDao()).thenReturn(mockAdditiveNameDao);
        when(mockDaoSession.getAdditiveDao()).thenReturn(mockAdditiveDao);
        when(mockAdditiveNameDao._queryAdditive_Names(any()))
            .thenReturn(Arrays.asList(ADDITIVE_NAME_1, ADDITIVE_NAME_2));

        mAdditive = new Additive();
    }

    @Test
    public void getNamesWithNullNamesAndNullDaoSession_throwsDaoException() {
        assertThrows(DaoException.class, () -> mAdditive.getNames());
    }

    @Test
    public void getNamesWithNullNamesAndNonNullDaoSession_setsNamesFromAdditiveNamesDao() {
        mAdditive.__setDaoSession(mockDaoSession);

        List<AdditiveName> names = mAdditive.getNames();

        assertThat(names).hasSize(2);
        assertThat(names.get(0).getName()).isEqualTo(ADDITIVE_NAME_NAME_1);
        assertThat(names.get(1).getName()).isEqualTo(ADDITIVE_NAME_NAME_2);
    }

    @Test
    public void deleteWithNullDao_throwsDaoException() {
        assertThrows(DaoException.class, () -> mAdditive.delete());
    }

    @Test
    public void deleteWithNonNullDao_callsDeleteOnDao() {
        mAdditive.__setDaoSession(mockDaoSession);

        mAdditive.delete();

        verify(mockAdditiveDao).delete(mAdditive);
    }

    @Test
    public void refreshWithNullDao_throwsDaoException() {
        assertThrows(DaoException.class, () -> mAdditive.refresh());
    }

    @Test
    public void refreshWithNonNullDao_callsRefreshOnDao() {
        mAdditive.__setDaoSession(mockDaoSession);

        mAdditive.refresh();

        verify(mockAdditiveDao).refresh(mAdditive);
    }

    @Test
    public void updateWithNullDao_throwsDaoException() {
        assertThrows(DaoException.class, () -> mAdditive.update());
    }

    @Test
    public void updateWithNonNullDao_callsUpdateOnDao() {
        mAdditive.__setDaoSession(mockDaoSession);

        mAdditive.update();

        verify(mockAdditiveDao).update(mAdditive);
    }
}

File: app/src/test/java/openfoodfacts/github/scrachx/openfood/models/entities/additive/AdditiveTest.kt
package openfoodfacts.github.scrachx.openfood.models.entities.additive

import com.google.common.truth.Truth.assertThat
import io.mockk.Runs
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.verify
import openfoodfacts.github.scrachx.openfood.models.DaoSession
import org.greenrobot.greendao.DaoException
import org.junit.Assert.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith

/**
 * Tests for [Additive]
 */
@ExtendWith(MockKExtension::class)
class AdditiveTest {
    @MockK
    private lateinit var mockDaoSession: DaoSession

    @MockK
    private lateinit var mockAdditiveDao: AdditiveDao

    @MockK
    private lateinit var mockAdditiveNameDao: AdditiveNameDao
    private lateinit var additive: Additive

    @BeforeEach
    fun beforeEach() {
        every { mockDaoSession.additiveNameDao } returns mockAdditiveNameDao
        every { mockDaoSession.additiveDao } returns mockAdditiveDao

        additive = Additive()
    }

    @Test
    fun `get names with null dao throws DaoException`() {
        assertThrows<DaoException> { additive.names }
    }

    @Test
    fun `get names with null names and non null dao sets names from AdditiveNamesDao`() {
        every { mockAdditiveNameDao._queryAdditive_Names(any()) }
            .returns(listOf(ADDITIVE_NAME_1, ADDITIVE_NAME_2))

        additive.__setDaoSession(mockDaoSession)

        val names = additive.names
        assertThat(names).hasSize(2)
        assertThat(names[0]!!.name).isEqualTo(ADDITIVE_NAME_NAME_1)
        assertThat(names[1]!!.name).isEqualTo(ADDITIVE_NAME_NAME_2)
    }

    @Test
    fun `delete with null dao throws DaoException`() {
        assertThrows<DaoException> { additive.delete() }
    }

    @Test
    fun `delete with non null dao calls delete on dao`() {
        every { mockAdditiveDao.delete(any()) } just Runs
        additive.__setDaoSession(mockDaoSession)
        additive.delete()
        verify { mockAdditiveDao.delete(additive) }
    }

    @Test
    fun `refresh with null dao throws DaoException`() {
        assertThrows(DaoException::class.java) { additive.refresh() }
    }

    @Test
    fun `refresh with non null dao calls refresh on dao`() {
        every { mockAdditiveDao.refresh(any()) } just Runs
        additive.__setDaoSession(mockDaoSession)
        additive.refresh()
        verify { mockAdditiveDao.refresh(additive) }
    }

    @Test
    fun `update with null dao throws DaoException`() {
        assertThrows<DaoException> { additive.update() }
    }

    @Test
    fun `update with non null dao calls update on dao`() {
        every { mockAdditiveDao.update(any()) } just Runs
        additive.__setDaoSession(mockDaoSession)
        additive.update()
        verify { mockAdditiveDao.update(additive) }
    }

    companion object {
        private const val ADDITIVE_NAME_NAME_1 = "AdditiveName"
        private const val ADDITIVE_NAME_NAME_2 = "AdditiveName2"
        private val ADDITIVE_NAME_1 = AdditiveName(ADDITIVE_NAME_NAME_1)
        private val ADDITIVE_NAME_2 = AdditiveName(ADDITIVE_NAME_NAME_2)
    }
}